### PR TITLE
801 reset password

### LINKED ISSUE #801: 🚀 [Feature] Reset password
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-12T10:04:41Z" title="Wednesday, June 12th 2019, 12:04:41 pm +02:00">Jun 12, 2019</time>_
_Closed <time datetime="2019-06-21T21:02:11Z" title="Friday, June 21st 2019, 11:02:11 pm +02:00">Jun 21, 2019</time>_
---

## :rocket: Feature

Reset your password and get a mail to login again.

#### User story
```gherkin
As a user
I want to have a button to send me a reset password mail
To be able to login again in case I forgot my password
```

### Design & Layout

Screenshot - 2019-06-12T120331 736

### PR DESCRIPTION
**Pull request by Tirokk**
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-15T09:49:21Z" title="Saturday, June 15th 2019, 11:49:21 am +02:00">Jun 15, 2019</time>_
_Merged <time datetime="2019-06-21T21:02:11Z" title="Friday, June 21st 2019, 11:02:11 pm +02:00">Jun 21, 2019</time>_
---

## 🍰 Pullrequest
Implement a basic password reset flow. I did not follow a security best practice here: You should ask the user for her username. That way, the account is safe even if an attacker has access to the email inbox. Reason for not implementing was that our users cannot choose their own slug at the moment. Therefore I believe we would make the whole flow pretty unusable if we ask for sth. that the user never changed and probably does not remember.

### Issues

- Close #801

### Todo

- [X] None

### CODE REVIEW

**Comment by Tirokk** [comment]
> [codecov[bot]](https://github.com/apps/codecov) **Authored by [codecov[bot]](https://github.com/apps/codecov)**
_<time datetime="2019-06-16T21:33:29Z" title="Sunday, June 16th 2019, 11:33:29 pm +02:00">Jun 16, 2019</time>_

---


# [Codecov](https://codecov.io/gh/Human-Connection/Human-Connection/pull/834?src=pr&el=h1) Report
> :exclamation: No coverage uploaded for pull request base (`master@53ce729`). [Click here to learn what that means](https://docs.codecov.io/docs/error-reference#section-missing-base-commit).
> The diff coverage is `n/a`.

**Comment by Tirokk** [comment]
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-19T23:35:22Z" title="Thursday, June 20th 2019, 1:35:22 am +02:00">Jun 20, 2019</time>_

---


This is still a little bit work-in-progress but I think this is the moment when I can ask you for a first review.

File: backend/src/schema/resolvers/passwordReset.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      }
+      return true
+    },
+    resetPassword: async (_, { email, code, newPassword }, { driver }) => {
+      const session = driver.session()
+      const stillValid = new Date()
+      stillValid.setDate(stillValid.getDate() - 1)
+      const newHashedPassword = await bcrypt.hashSync(newPassword, 10)
+      const cypher = `
+      MATCH (pr:PasswordReset {code: $code})
```
Comment (Tirokk) [inline backend/src/schema/resolvers/passwordReset.js:58]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-06-20T07:00:53Z" title="Thursday, June 20th 2019, 9:00:53 am +02:00">Jun 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,72 @@
+import uuid from 'uuid/v4'
+import bcrypt from 'bcryptjs'
+import CONFIG from '../../config'
+import nodemailer from 'nodemailer'
+import { resetPasswordMail, wrongAccountMail } from './passwordReset/emailTemplates'
+
+const transporter = () => {
+  const { SMTP_HOST: host, SMTP_PORT: port, SMTP_USERNAME: user, SMTP_PASSWORD: pass } = CONFIG
+  const configs = {
+    host,
+    port,
+    ignoreTLS: true,
+    secure: false, // true for 465, false for other ports
+  }
+  if (user && pass) {
+    configs.auth = { user, pass }
+  }
+  return nodemailer.createTransport(configs)
+}
+
+export async function createPasswordReset(options) {
+  const { driver, code, email, issuedAt = new Date() } = options
+  const session = driver.session()
+  const cypher = `
+      MATCH (u:User) WHERE u.email = $email
+      CREATE(pr:PasswordReset {code: $code, issuedAt: datetime($issuedAt), usedAt: NULL})
+      MERGE (u)-[:REQUESTED]->(pr)
+      RETURN u
+      `
+  const transactionRes = await session.run(cypher, {
+    issuedAt: issuedAt.toISOString(),
+    code,
+    email,
+  })
+  const users = transactionRes.records.map(record => record.get('u'))
+  session.close()
+  return users
+}
+
+export default {
+  Mutation: {
+    requestPasswordReset: async (_, { email }, { driver }) => {
+      const code = uuid().substring(0, 6)
+      const [user] = await createPasswordReset({ driver, code, email })
+      if (CONFIG.SMTP_HOST && CONFIG.SMTP_PORT) {
+        const name = (user && user.name) || ''
+        const mailTemplate = user ? resetPasswordMail : wrongAccountMail
+        await transporter().sendMail(mailTemplate({ email, code, name }))
+      }
+      return true
+    },
+    resetPassword: async (_, { email, code, newPassword }, { driver }) => {
+      const session = driver.session()
+      const stillValid = new Date()
+      stillValid.setDate(stillValid.getDate() - 1)
+      const newHashedPassword = await bcrypt.hashSync(newPassword, 10)
+      const cypher = `
+      MATCH (r:PasswordReset {code: $code})
```


Would be nice to name the `PasswordReset` not `r`, but `pr` as above. @roschaefer

File: backend/src/schema/resolvers/passwordReset.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      RETURN pr
+      `
+      let transactionRes = await session.run(cypher, { stillValid, email, code, newHashedPassword })
+      const [reset] = transactionRes.records.map(record => record.get('pr'))
+      const result = !!(reset && reset.properties.usedAt)
+      session.close()
+      return result
+    },
+  },
+}
```
Comment (Tirokk) [inline backend/src/schema/resolvers/passwordReset.js:72]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-06-20T07:05:11Z" title="Thursday, June 20th 2019, 9:05:11 am +02:00">Jun 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,72 @@
+import uuid from 'uuid/v4'
+import bcrypt from 'bcryptjs'
+import CONFIG from '../../config'
+import nodemailer from 'nodemailer'
+import { resetPasswordMail, wrongAccountMail } from './passwordReset/emailTemplates'
+
+const transporter = () => {
+  const { SMTP_HOST: host, SMTP_PORT: port, SMTP_USERNAME: user, SMTP_PASSWORD: pass } = CONFIG
+  const configs = {
+    host,
+    port,
+    ignoreTLS: true,
+    secure: false, // true for 465, false for other ports
+  }
+  if (user && pass) {
+    configs.auth = { user, pass }
+  }
+  return nodemailer.createTransport(configs)
+}
+
+export async function createPasswordReset(options) {
+  const { driver, code, email, issuedAt = new Date() } = options
+  const session = driver.session()
+  const cypher = `
+      MATCH (u:User) WHERE u.email = $email
+      CREATE(pr:PasswordReset {code: $code, issuedAt: datetime($issuedAt), usedAt: NULL})
+      MERGE (u)-[:REQUESTED]->(pr)
+      RETURN u
+      `
+  const transactionRes = await session.run(cypher, {
+    issuedAt: issuedAt.toISOString(),
+    code,
+    email,
+  })
+  const users = transactionRes.records.map(record => record.get('u'))
+  session.close()
+  return users
+}
+
+export default {
+  Mutation: {
+    requestPasswordReset: async (_, { email }, { driver }) => {
+      const code = uuid().substring(0, 6)
+      const [user] = await createPasswordReset({ driver, code, email })
+      if (CONFIG.SMTP_HOST && CONFIG.SMTP_PORT) {
+        const name = (user && user.name) || ''
+        const mailTemplate = user ? resetPasswordMail : wrongAccountMail
+        await transporter().sendMail(mailTemplate({ email, code, name }))
+      }
+      return true
+    },
+    resetPassword: async (_, { email, code, newPassword }, { driver }) => {
+      const session = driver.session()
+      const stillValid = new Date()
+      stillValid.setDate(stillValid.getDate() - 1)
+      const newHashedPassword = await bcrypt.hashSync(newPassword, 10)
+      const cypher = `
+      MATCH (r:PasswordReset {code: $code})
+      MATCH (u:User {email: $email})-[:REQUESTED]->(r)
+      WHERE duration.between(r.issuedAt, datetime()).days <= 0 AND r.usedAt IS NULL
+      SET r.usedAt = datetime()
+      SET u.password = $newHashedPassword
+      RETURN r
+      `
+      let transactionRes = await session.run(cypher, { stillValid, email, code, newHashedPassword })
+      const [reset] = transactionRes.records.map(record => record.get('r'))
+      const result = !!(reset && reset.properties.usedAt)
+      session.close()
+      return result
+    },
+  },
+}
```


So, the `PasswordReset ` node will be never deleted?
Do I understand right? @roschaefer

File: backend/src/schema/resolvers/passwordReset.spec.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+      password: '1234',
+    })
+  })
+
+  afterEach(async () => {
+    await factory.cleanDatabase()
+  })
+
+  describe('requestPasswordReset', () => {
+    const mutation = `mutation($email: String!) { requestPasswordReset(email: $email) }`
```
Comment (Tirokk) [inline backend/src/schema/resolvers/passwordReset.spec.js:34]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-06-20T07:09:29Z" title="Thursday, June 20th 2019, 9:09:29 am +02:00">Jun 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,180 @@
+import { GraphQLClient } from 'graphql-request'
+import Factory from '../../seed/factories'
+import { host } from '../../jest/helpers'
+import { getDriver } from '../../bootstrap/neo4j'
+import { createPasswordReset } from './passwordReset'
+
+const factory = Factory()
+let client
+const driver = getDriver()
+
+const getAllPasswordResets = async () => {
+  const session = driver.session()
+  let transactionRes = await session.run('MATCH (r:PasswordReset) RETURN r')
+  const resets = transactionRes.records.map(record => record.get('r'))
+  session.close()
+  return resets
+}
+
+describe('passwordReset', () => {
+  beforeEach(async () => {
+    client = new GraphQLClient(host)
+    await factory.create('User', {
+      email: 'user@example.org',
+      role: 'user',
+      password: '1234',
+    })
+  })
+
+  afterEach(async () => {
+    await factory.cleanDatabase()
+  })
+
+  describe('requestPasswordReset', () => {
+    const mutation = `mutation($email: String!) { requestPasswordReset(email: $email) }`
```


If you use `gql` by `import gql from 'graphql-tag‘` it looks nicer.

File: backend/src/schema/resolvers/passwordReset/emailTemplates.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+If you did not request a password reset, please ignore this email or contact
+support if you have questions:
+
+${supportUrl}
+
+Thanks,
+The Human Connection Team
+
+If you're having trouble with the link above, you can manually copy and
```
Comment (Tirokk) [inline backend/src/schema/resolvers/passwordReset/emailTemplates.js:37]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-06-20T07:46:30Z" title="Thursday, June 20th 2019, 9:46:30 am +02:00">Jun 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,85 @@
+import CONFIG from '../../../config'
+
+export const from = '"Human Connection" <info@human-connection.org>'
+
+export const resetPasswordMail = options => {
+  const {
+    name,
+    email,
+    code,
+    subject = 'Use this link to reset your password. The link is only valid for 24 hours.',
+    supportUrl = 'https://human-connection.org/en/contact/',
+  } = options
+  const actionUrl = new URL('/password-reset/change-password', CONFIG.CLIENT_URI)
+  actionUrl.searchParams.set('code', code)
+  actionUrl.searchParams.set('email', email)
+
+  return {
+    to: email,
+    subject,
+    text: `
+Hi ${name}!
+
+You recently requested to reset your password for your Human Connection account.
+Use the button below to reset it. This password reset is only valid for the next
+24 hours.
+
+${actionUrl}
+
+If you did not request a password reset, please ignore this email or contact
+support if you have questions:
+
+${supportUrl}
+
+Thanks,
+The Human Connection Team
+
+If you’re having trouble with the link above, you can manually copy and
```


Please don’t use `’` here, instead use `'` I think. @roschaefer

File: webapp/locales/de.json [hunk 5]
```diff
@@ -8,11 +8,32 @@
...
+    "form": {
+      "description": "Eine Mail zum Zurücksetzen des Passworts wird an die angegebene E-Mail Adresse geschickt.",
+      "submit": "Email anfordern",
+      "submitted": "Eine E-Mail mit weiteren Instruktionen wurde verschickt an <b>{email}</b>"
+    }
+  },
+  "verify-code": {
+    "form": {
+      "code": "Code eingeben",
+      "description": "Öffne dein E-Mail Postfach und gib den Code ein, den wir geschickt haben.",
```
Comment (Tirokk) [inline webapp/locales/de.json:28]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-06-20T07:57:33Z" title="Thursday, June 20th 2019, 9:57:33 am +02:00">Jun 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -8,11 +8,32 @@
     "logout": "Ausloggen",
     "email": "Deine E-Mail",
     "password": "Dein Passwort",
+    "forgotPassword": "Passwort vergessen?",
     "moreInfo": "Was ist Human Connection?",
     "moreInfoURL": "https://human-connection.org",
     "moreInfoHint": "zur Präsentationsseite",
     "hello": "Hallo"
   },
+  "password-reset": {
+    "title": "Passwort zurücksetzen",
+    "form": {
+      "description": "Eine Mail zum Zurücksetzen des Passworts wird an die angegebene E-Mail Adresse geschickt.",
+      "submit": "Email anfordern",
+      "submitted": "Eine E-Mail mit weiteren Instruktionen wurde verschickt an <b>{email}</b>"
+    }
+  },
+  "verify-code": {
+    "form": {
+      "code": "Code eingeben",
+      "description": "Öffne Deine E-Mail Postfach und gib den Code ein, den wir geschickt haben.",
```


`Öffne Dein E-Mail Postfach …`

File: webapp/locales/de.json [hunk 6]
```diff
@@ -8,11 +8,32 @@
...
+      "next": "Weiter",
+      "change-password":{
+        "success": "Änderung des Passworts war erfolgreich!",
```
Comment (Tirokk) [inline webapp/locales/de.json:31]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-06-20T08:00:03Z" title="Thursday, June 20th 2019, 10:00:03 am +02:00">Jun 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -8,11 +8,32 @@
     "logout": "Ausloggen",
     "email": "Deine E-Mail",
     "password": "Dein Passwort",
+    "forgotPassword": "Passwort vergessen?",
     "moreInfo": "Was ist Human Connection?",
     "moreInfoURL": "https://human-connection.org",
     "moreInfoHint": "zur Präsentationsseite",
     "hello": "Hallo"
   },
+  "password-reset": {
+    "title": "Passwort zurücksetzen",
+    "form": {
+      "description": "Eine Mail zum Zurücksetzen des Passworts wird an die angegebene E-Mail Adresse geschickt.",
+      "submit": "Email anfordern",
+      "submitted": "Eine E-Mail mit weiteren Instruktionen wurde verschickt an <b>{email}</b>"
+    }
+  },
+  "verify-code": {
+    "form": {
+      "code": "Code eingeben",
+      "description": "Öffne Deine E-Mail Postfach und gib den Code ein, den wir geschickt haben.",
+      "next": "Weiter",
+      "change-password":{
+        "success": "Änderung des Passworts war erfolgreich",
```


`… war erfolgreich!` we usually put a `!` behind success texts I have the impression.

File: webapp/locales/en.json [hunk 7]
```diff
@@ -8,11 +8,32 @@
...
+      "submitted": "A mail with further instruction has been sent to <b>{email}</b>"
+    }
+  },
+  "verify-code": {
+    "form": {
+      "code": "Enter your code",
+      "description": "Open your inbox and enter the code that we've sent to you.",
+      "next": "Continue",
+      "change-password": {
+        "success": "Changing your password was successful!",
```
Comment (Tirokk) [inline webapp/locales/en.json:31]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-06-20T08:01:39Z" title="Thursday, June 20th 2019, 10:01:39 am +02:00">Jun 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -8,11 +8,32 @@
     "logout": "Logout",
     "email": "Your Email",
     "password": "Your Password",
+    "forgotPassword": "Forgot Password?",
     "moreInfo": "What is Human Connection?",
     "moreInfoURL": "https://human-connection.org/en/",
     "moreInfoHint": "to the presentation page",
     "hello": "Hello"
   },
+  "password-reset": {
+    "title": "Reset your password",
+    "form": {
+      "description": "A password reset email will be sent to the given email address.",
+      "submit": "Request email",
+      "submitted": "A mail with further instruction has been sent to <b>{email}</b>"
+    }
+  },
+  "verify-code": {
+    "form": {
+      "code": "Enter your code",
+      "description": "Open your inbox and enter the code that we've sent to you.",
+      "next": "Continue",
+      "change-password": {
+        "success": "Changing your password was successful",
```


`… successful!` we usually put a ! behind success texts I have the impression.

File: backend/src/schema/resolvers/passwordReset.js (hunk 2 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/passwordReset.js:72]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-20T13:21:53Z" title="Thursday, June 20th 2019, 3:21:53 pm +02:00">Jun 20, 2019</time>_

---


yes

File: backend/src/schema/resolvers/passwordReset.spec.js (hunk 3 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/passwordReset.spec.js:34]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-20T13:22:46Z" title="Thursday, June 20th 2019, 3:22:46 pm +02:00">Jun 20, 2019</time>_

---


In tests? I usually don't do that - by looking nicer you mean the editor is doing syntax highlighting?

File: backend/src/schema/resolvers/passwordReset/emailTemplates.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  actionUrl.searchParams.set('email', email)
+
+  return {
+    to: email,
+    subject,
+    text: `
+Hi ${name}!
+
+You recently requested to reset your password for your Human Connection account.
+Use the link below to reset it. This password reset is only valid for the next
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply backend/src/schema/resolvers/passwordReset/emailTemplates.js:24]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-20T13:23:09Z" title="Thursday, June 20th 2019, 3:23:09 pm +02:00">Jun 20, 2019</time>_

---


Yes, you're right

File: webapp/locales/de.json (hunk 6 above)
Comment (Tirokk) [reply webapp/locales/de.json:31]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-20T13:24:04Z" title="Thursday, June 20th 2019, 3:24:04 pm +02:00">Jun 20, 2019</time>_

---


:+1:

**Comment by Tirokk** [comment]
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-20T13:26:13Z" title="Thursday, June 20th 2019, 3:26:13 pm +02:00">Jun 20, 2019</time>_

---


@Tirokk does `yarn install --ignore-engines` resolve the problem? I have never seen that output on my machine and believe it's unrelated.

File: backend/src/schema/resolvers/passwordReset.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    },
+    resetPassword: async (_, { email, code, newPassword }, { driver }) => {
+      const session = driver.session()
+      const stillValid = new Date()
+      stillValid.setDate(stillValid.getDate() - 1)
+      const newHashedPassword = await bcrypt.hashSync(newPassword, 10)
+      const cypher = `
+      MATCH (pr:PasswordReset {code: $code})
+      MATCH (u:User {email: $email})-[:REQUESTED]->(pr)
+      WHERE duration.between(pr.issuedAt, datetime()).days <= 0 AND pr.usedAt IS NULL
```
Comment (Tirokk) [inline backend/src/schema/resolvers/passwordReset.js:60]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-20T20:33:54Z" title="Thursday, June 20th 2019, 10:33:54 pm +02:00">Jun 20, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,72 @@
+import uuid from 'uuid/v4'
+import bcrypt from 'bcryptjs'
+import CONFIG from '../../config'
+import nodemailer from 'nodemailer'
+import { resetPasswordMail, wrongAccountMail } from './passwordReset/emailTemplates'
+
+const transporter = () => {
+  const configs = {
+    host: CONFIG.SMTP_HOST,
+    port: CONFIG.SMTP_PORT,
+    ignoreTLS: CONFIG.SMTP_IGNORE_TLS,
+    secure: false, // true for 465, false for other ports
+  }
+  const { SMTP_USERNAME: user, SMTP_PASSWORD: pass } = CONFIG
+  if (user && pass) {
+    configs.auth = { user, pass }
+  }
+  return nodemailer.createTransport(configs)
+}
+
+export async function createPasswordReset(options) {
+  const { driver, code, email, issuedAt = new Date() } = options
+  const session = driver.session()
+  const cypher = `
+      MATCH (u:User) WHERE u.email = $email
+      CREATE(pr:PasswordReset {code: $code, issuedAt: datetime($issuedAt), usedAt: NULL})
+      MERGE (u)-[:REQUESTED]->(pr)
+      RETURN u
+      `
+  const transactionRes = await session.run(cypher, {
+    issuedAt: issuedAt.toISOString(),
+    code,
+    email,
+  })
+  const users = transactionRes.records.map(record => record.get('u'))
+  session.close()
+  return users
+}
+
+export default {
+  Mutation: {
+    requestPasswordReset: async (_, { email }, { driver }) => {
+      const code = uuid().substring(0, 6)
+      const [user] = await createPasswordReset({ driver, code, email })
+      if (CONFIG.SMTP_HOST && CONFIG.SMTP_PORT) {
+        const name = (user && user.name) || ''
+        const mailTemplate = user ? resetPasswordMail : wrongAccountMail
+        await transporter().sendMail(mailTemplate({ email, code, name }))
+      }
+      return true
+    },
+    resetPassword: async (_, { email, code, newPassword }, { driver }) => {
+      const session = driver.session()
+      const stillValid = new Date()
+      stillValid.setDate(stillValid.getDate() - 1)
+      const newHashedPassword = await bcrypt.hashSync(newPassword, 10)
+      const cypher = `
+      MATCH (pr:PasswordReset {code: $code})
+      MATCH (u:User {email: $email})-[:REQUESTED]->(pr)
+      WHERE duration.between(pr.issuedAt, datetime()).days <= 0 AND pr.usedAt IS NULL
```


Maybe I'm missing something... the duration between when the password reset was issued and now in days is `less` than or equal to 0? shouldn't it be 1??

Comment (Tirokk) [reply backend/src/schema/resolvers/passwordReset.js:60]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-21T12:26:12Z" title="Friday, June 21st 2019, 2:26:12 pm +02:00">Jun 21, 2019</time>_

---


The duration will be a full day. So `<=1` would mean 1 day + 23 hours + 59 minutes + 59 seconds.

File: backend/src/schema/resolvers/passwordReset.spec.js (hunk 3 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/passwordReset.spec.js:34]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-06-21T12:33:32Z" title="Friday, June 21st 2019, 2:33:32 pm +02:00">Jun 21, 2019</time>_

---


In tests! Yes!
Test should be good readable as well, or?

Comment (Tirokk) [reply backend/src/schema/resolvers/passwordReset.spec.js:34]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-21T14:49:17Z" title="Friday, June 21st 2019, 4:49:17 pm +02:00">Jun 21, 2019</time>_

---


I think the point is that @roschaefer doesn't use VS code

File: backend/src/schema/resolvers/passwordReset.js (hunk 9 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/passwordReset.js:60]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-21T16:31:45Z" title="Friday, June 21st 2019, 6:31:45 pm +02:00">Jun 21, 2019</time>_

---


Sorry, maybe I am being dense, but how is less than or equal to 1 mean 1+?

**Review by Mogge — CHANGES_REQUESTED** [review]
Hey @roschaefer !!! 😎
That’s really a masterpiece. Thank you for all this work. 🚀🚀💫
This will take us off …

Unfortunately we have some console output in the frontend tests and on the bowser console:

Bildschirmfoto 2019-06-20 um 09 55 48

Bildschirmfoto 2019-06-20 um 09 35 37

Does this belong to you?
I think the browser console output belongs to the deletion of the users account!?

I tried the reset password out:
- I change the `Jenny Rostocks` e-mail to mine, before.
- But I didn’t received an e-mail on my account.

Should that work on local installation (without Docker)?

To get around that issue I:

- Made a console output of the generated code.
- This did work with me, as I put the code in the appropriate input field. 👋🏼

**Review by Mogge — APPROVED** [review]
Looks great @roschaefer!
I had a doubt that I asked about :point_down: if you could respond
Also, one thing I really liked about this `PR` is your use of destructuring to pull out values from `this.form`... I will refactoring my code to use this as well :smiley:

**Review by Mogge — APPROVED** [review]
Fantastic and awesome job @roschaefer !!!
Your hard work was worth it!

You got it. 😍

Gandalfs stick

I didn't get the e-mail to send, not even with Docker.
But this is certainly a weak point of my own abilities.... 😓

**Pull request closed without merging**